In [1]:
import re
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer

In [2]:
#lets start this with a smaller data:

data= 'hello everybody!. Well-come to NLP tutorial. We are going to see the simple text preprocessing. I am going \
to use stemmer & lemmatization from nltk library.'

In [3]:
sents= nltk.sent_tokenize(data)

In [4]:
#lets clean the data :

corpus=[]

for i in range(len(sents)):
    review= re.sub('[^a-zA-Z]',' ', sents[i])
    review= review.lower()
    corpus.append(review)

In [5]:
corpus

['hello everybody  ',
 'well come to nlp tutorial ',
 'we are going to see the simple text preprocessing ',
 'i am going to use stemmer   lemmatization from nltk library ']

In [6]:
#lets apply TF-IDF to the above corpus

tf_idf = TfidfVectorizer()

tf_idf_vects= tf_idf.fit_transform(corpus)

In [7]:
tf_idf.get_feature_names_out()

array(['am', 'are', 'come', 'everybody', 'from', 'going', 'hello',
       'lemmatization', 'library', 'nlp', 'nltk', 'preprocessing', 'see',
       'simple', 'stemmer', 'text', 'the', 'to', 'tutorial', 'use', 'we',
       'well'], dtype=object)

In [8]:
tf_idf.vocabulary_

{'hello': 6,
 'everybody': 3,
 'well': 21,
 'come': 2,
 'to': 17,
 'nlp': 9,
 'tutorial': 18,
 'we': 20,
 'are': 1,
 'going': 5,
 'see': 12,
 'the': 16,
 'simple': 13,
 'text': 15,
 'preprocessing': 11,
 'am': 0,
 'use': 19,
 'stemmer': 14,
 'lemmatization': 7,
 'from': 4,
 'nltk': 10,
 'library': 8}

In [9]:
# get the transformed vector:

tf_idf_vects.toarray()

array([[0.        , 0.        , 0.        , 0.70710678, 0.        ,
        0.        , 0.70710678, 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        ],
       [0.        , 0.        , 0.47633035, 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.47633035,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.30403549, 0.47633035, 0.        ,
        0.        , 0.47633035],
       [0.        , 0.35291425, 0.        , 0.        , 0.        ,
        0.27824164, 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.35291425, 0.35291425, 0.35291425, 0.        ,
        0.35291425, 0.35291425, 0.22526059, 0.        , 0.        ,
        0.35291425, 0.        ],
       [0.35291425, 0.        , 0.        , 0.        , 0.35291425,
        0.2782416

In [29]:
# lets chekc the size fo the vector

tf_idf_vects.toarray().shape

(4, 22)

Note: 4 sentences and each sentence has 22 vocab_size

In [10]:
# if you want the 1st sentence vector:

tf_idf_vects[0].toarray()

array([[0.        , 0.        , 0.        , 0.70710678, 0.        ,
        0.        , 0.70710678, 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        ]])

In [11]:
# lets apply TF-IDF on a larger data:

paragraph =  open('C:/Users/acreddy/Desktop/spacy_ds/nlp.txt').read()

In [12]:
paragraph

'I have three visions for India. In 3000 years of our history, people from all over \nthe world have come and invaded us, captured our lands, conquered our minds. \nFrom Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,\nthe French, the Dutch, all of them came and looted us, took over what was ours. \nYet we have not done this to any other nation. We have not conquered anyone. \nWe have not grabbed their land, their culture, \ntheir history and tried to enforce our way of life on them. \nWhy? Because we respect the freedom of others.That is why my \nfirst vision is that of freedom. I believe that India got its first vision of \nthis in 1857, when we started the War of Independence. It is this freedom that\nwe must protect and nurture and build on. If we are not free, no one will respect us.\nMy second vision for India’s development. For fifty years we have been a developing nation.\nIt is time we see ourselves as a developed nation. We are among the t

In [13]:
# breakdown in to sentences:

sentences= nltk.sent_tokenize(paragraph)

In [14]:
sentences

['I have three visions for India.',
 'In 3000 years of our history, people from all over \nthe world have come and invaded us, captured our lands, conquered our minds.',
 'From Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,\nthe French, the Dutch, all of them came and looted us, took over what was ours.',
 'Yet we have not done this to any other nation.',
 'We have not conquered anyone.',
 'We have not grabbed their land, their culture, \ntheir history and tried to enforce our way of life on them.',
 'Why?',
 'Because we respect the freedom of others.That is why my \nfirst vision is that of freedom.',
 'I believe that India got its first vision of \nthis in 1857, when we started the War of Independence.',
 'It is this freedom that\nwe must protect and nurture and build on.',
 'If we are not free, no one will respect us.',
 'My second vision for India’s development.',
 'For fifty years we have been a developing nation.',
 'It is time we see ourselve

In [20]:
# clean the data:

corpus1= []

for i in range(len(sentences)):
    review= re.sub('[^a-zA-Z0-9]',' ', sentences[i])
    review= review.lower()
    corpus1.append(review)

In [21]:
corpus1

['i have three visions for india ',
 'in 3000 years of our history  people from all over  the world have come and invaded us  captured our lands  conquered our minds ',
 'from alexander onwards  the greeks  the turks  the moguls  the portuguese  the british  the french  the dutch  all of them came and looted us  took over what was ours ',
 'yet we have not done this to any other nation ',
 'we have not conquered anyone ',
 'we have not grabbed their land  their culture   their history and tried to enforce our way of life on them ',
 'why ',
 'because we respect the freedom of others that is why my  first vision is that of freedom ',
 'i believe that india got its first vision of  this in 1857  when we started the war of independence ',
 'it is this freedom that we must protect and nurture and build on ',
 'if we are not free  no one will respect us ',
 'my second vision for india   s development ',
 'for fifty years we have been a developing nation ',
 'it is time we see ourselves as a

In [22]:
# let sapply TF-IDF:

tf_idf= TfidfVectorizer()

tf_idf_vectors= tf_idf.fit_transform(corpus1)

In [24]:
tf_idf.get_feature_names_out()

array(['10', '1857', '3000', 'achievements', 'alexander', 'all', 'also',
       'among', 'an', 'and', 'any', 'anyone', 'are', 'areas', 'as',
       'assured', 'be', 'because', 'been', 'being', 'believe', 'both',
       'brahm', 'british', 'build', 'but', 'came', 'captured', 'career',
       'closely', 'come', 'confidence', 'conquered', 'consider',
       'culture', 'dept', 'developed', 'developing', 'development',
       'dhawan', 'done', 'dr', 'dutch', 'economic', 'enforce', 'falling',
       'father', 'fifty', 'first', 'for', 'fortune', 'four', 'free',
       'freedom', 'french', 'from', 'gdp', 'globally', 'go', 'good',
       'got', 'grabbed', 'great', 'greeks', 'growth', 'hand', 'have',
       'him', 'history', 'if', 'in', 'incorrect', 'independence', 'india',
       'invaded', 'is', 'isn', 'it', 'its', 'lack', 'land', 'lands',
       'levels', 'life', 'looted', 'lucky', 'material', 'milestones',
       'military', 'minds', 'moguls', 'most', 'must', 'my', 'nation',
       'nations'

In [25]:
tf_idf.vocabulary_

{'have': 66,
 'three': 144,
 'visions': 157,
 'for': 49,
 'india': 73,
 'in': 70,
 '3000': 2,
 'years': 170,
 'of': 100,
 'our': 108,
 'history': 68,
 'people': 112,
 'from': 55,
 'all': 5,
 'over': 111,
 'the': 139,
 'world': 169,
 'come': 30,
 'and': 9,
 'invaded': 74,
 'us': 154,
 'captured': 27,
 'lands': 81,
 'conquered': 32,
 'minds': 89,
 'alexander': 4,
 'onwards': 104,
 'greeks': 63,
 'turks': 151,
 'moguls': 90,
 'portuguese': 114,
 'british': 23,
 'french': 54,
 'dutch': 42,
 'them': 141,
 'came': 26,
 'looted': 84,
 'took': 148,
 'what': 162,
 'was': 159,
 'ours': 109,
 'yet': 171,
 'we': 161,
 'not': 97,
 'done': 40,
 'this': 143,
 'to': 146,
 'any': 10,
 'other': 106,
 'nation': 94,
 'anyone': 11,
 'grabbed': 61,
 'their': 140,
 'land': 80,
 'culture': 34,
 'tried': 150,
 'enforce': 44,
 'way': 160,
 'life': 83,
 'on': 101,
 'why': 165,
 'because': 17,
 'respect': 123,
 'freedom': 53,
 'others': 107,
 'that': 138,
 'is': 75,
 'my': 93,
 'first': 48,
 'vision': 156,
 'beli

In [26]:
#lets get the transformed vectors:

tf_idf_vectors.toarray()

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.22629718, ..., 0.17133386, 0.20197552,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [27]:
tf_idf_vectors.toarray().shape

(31, 172)

Note: 31 sentences and each sentence having 172 vocab_size

In [36]:
#lets apply ngram and max_frequency(max_features value gives the minimum those many times repeated values only. i,e. if i give max_features=2 then it prints only the words who repeated more than 2 times in corpus 

tf_idf= TfidfVectorizer(ngram_range=(1,2),max_features=2)

tf_id_vect= tf_idf.fit_transform(corpus1)

In [37]:
# it gives only the minimum 2 times repeated words only in the corpus:

tf_idf.get_feature_names_out()

array(['of', 'the'], dtype=object)

In [38]:
# since it has only 2 words repeated 2 or more times:

tf_idf.vocabulary_

{'of': 0, 'the': 1}